### ISSUE TITLE
Deadlock when closing channels

### ISSUE THREAD

**Issue by nvx** [streadway/amqp#147]
Out of habit, I am using "defer chan.Close()" after obtaining a channel.

Channels are returned from a separate goroutine that acts as a channel factory (handling reconnection all in one place). When I send a signal to the context to shutdown everything gracefully, I find my workers hung in chan.Close(). My suspicion is that conn.Close() is being called first (or simultaneously) and either conn.Close() followed by chan.Close() hangs, or conn.Close() running simultaneously with chan.Close() causes the latter to hang.

As a quick fix I just removed the chan.Close() and everything seems to be fine as conn.Close() appears to clean up channels anyway, but I feel like this should be fixed.

**Comment by domashevskiy** [comment]
Sample code. It hangs in most cases. If you build it with "-race" flag you'll see data race details.
```go
package main

import (
	"fmt"
	"github.com/streadway/amqp"
	"log"
	"runtime"
	"sync"
	"time"
)

func init() {
	runtime.GOMAXPROCS(runtime.NumCPU())
}

func createChannels(conn *amqp.Connection, wg *sync.WaitGroup) {
	ch, err := conn.Channel()
	if err != nil {
		panic(err)
	}

	time.Sleep(time.Second)
	fmt.Println("Chan close")

	err = ch.Close()

	wg.Done()
}

func channels() {
	conn, err := amqp.Dial("amqp://login:pass@localhost:5672/")
	if err != nil {
		log.Fatalf("connection.open: %s", err)
	}

	wg := new(sync.WaitGroup)

	for i := 0; i < 8; i++ {
		wg.Add(1)
		go createChannels(conn, wg)
	}
	time.Sleep(time.Second)
	fmt.Println("Conn close")
	conn.Close()

	wg.Wait()
}

func main() {
	channels()
}
```

**Comment by domashevskiy** [comment]
It hangs in channel.go's call function. It waits infinitely to receive something from me.rpc. Maybe you should close rpc channel on shutdown? 
In my experiments adding close(me.rpc) in shutdown method of a channel helped.
```go
// Performs a request/response call for when the message is not NoWait and is
// specified as Synchronous.
func (me *Channel) call(req message, res ...message) error {
	if err := me.send(me, req); err != nil {
		return err
	}

	if req.wait() {
		select {
		case e := <-me.errors:
			return e

		case msg := <-me.rpc:
			if msg != nil {
				for _, try := range res {
					if reflect.TypeOf(msg) == reflect.TypeOf(try) {
						// *res = *msg
						vres := reflect.ValueOf(try).Elem()
						vmsg := reflect.ValueOf(msg).Elem()
						vres.Set(vmsg)
						return nil
					}
				}
				return ErrCommandInvalid
			} else {
				// RPC channel has been closed without an error, likely due to a hard
				// error on the Connection.  This indicates we have already been
				// shutdown and if were waiting, will have returned from the errors chan.
				return ErrClosed
			}
		}
	}

	return nil
}
```

**Comment by nvx** [comment]
Cheers for the sample @domashevskiy - was in another frame of mind deep in another project earlier when I logged this and reproducing it simply was rather beyond me at the time haha.

I tried go run -race and it actually failed to report anything in my specific setup which was infuriating to say the least.

**Comment by joliver** [comment]
I just bumped into this one as well. I'm cancelling a consumer on a failed channel using `channel.Cancel(...)` and the channel waits forever in on a `select` statement here:
https://github.com/streadway/amqp/blob/master/channel.go#L149

I'm experienced this on an earlier version of the code so I pulled the latest commit (f4879ba28fffbb576743b03622a9ff20461826b2) and it's still there.

Adding a close(me.rpc) during the channel shutdown solves the problem.
